// nourishnet-app/src/screens/LoginScreen.tsx
import React, { useState } from 'react'
import { View, Text, StyleSheet, TouchableOpacity, Alert } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from '../store/store'
import { loginUser, registerUser } from '../store/slices/authSlice'

const LoginScreen = () => {
  const [isLoading, setIsLoading] = useState(false)
  const dispatch = useDispatch()
  const { error } = useSelector((state: RootState) => state.auth)

  const handleDemoLogin = async () => {
    try {
      setIsLoading(true)
      
      // Create a demo user
      const userData = {
        id: 'demo-consumer-id',
        email: 'demo@nourishnet.com',
        name: 'Demo Consumer',
        role: 'consumer' as const,
        phone: '+91 98765 43210',
        address: {
          street: '456 Consumer Street',
          city: 'Mumbai',
          state: 'Maharashtra',
          pincode: '400001',
          coordinates: {
            lat: 19.0760,
            lng: 72.8777
          }
        },
        isActive: true,
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString()
      }
      
      // Direct login without API call for demo
      dispatch(loginUser(userData))
      Alert.alert('Success', 'Demo login successful!')
    } catch (error: any) {
      console.error('Login error:', error)
      Alert.alert('Error', error.message || 'Login failed')
    } finally {
      setIsLoading(false)
    }
  }

  const handleAnonymousLogin = () => {
    // Create an anonymous user
    const anonymousUser = {
      id: 'anonymous-user-id',
      email: 'anonymous@nourishnet.com',
      name: 'Anonymous User',
      role: 'consumer' as const,
      phone: '',
      address: {
        street: '',
        city: '',
        state: '',
        pincode: '',
        coordinates: {
          lat: 0,
          lng: 0
        }
      },
      isActive: true,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    }
    
    dispatch(loginUser(anonymousUser))
    Alert.alert('Success', 'Proceeding anonymously!')
  }

  return (
    <View style={styles.container}>
      <View style={styles.content}>
        <Text style={styles.title}>NourishNet</Text>
        <Text style={styles.subtitle}>Your Tiffin Service Companion</Text>
        
        <TouchableOpacity
          style={[styles.button, styles.primaryButton, isLoading && styles.buttonDisabled]}
          onPress={handleDemoLogin}
          disabled={isLoading}
        >
          <Text style={styles.buttonText}>
            {isLoading ? 'Signing in...' : 'Demo Login (Consumer)'}
          </Text>
        </TouchableOpacity>
        
        <TouchableOpacity
          style={[styles.button, styles.secondaryButton]}
          onPress={handleAnonymousLogin}
        >
          <Text style={[styles.buttonText, styles.secondaryButtonText]}>
            Continue Anonymously
          </Text>
        </TouchableOpacity>
        
        {error && (
          <Text style={styles.errorText}>{error}</Text>
        )}
        
        <Text style={styles.demoText}>
          This is a simplified MVP version for demonstration purposes
        </Text>
      </View>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f8fafc',
  },
  content: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  title: {
    fontSize: 32,
    fontWeight: 'bold',
    color: '#0ea5e9',
    marginBottom: 8,
  },
  subtitle: {
    fontSize: 16,
    color: '#64748b',
    marginBottom: 40,
    textAlign: 'center',
  },
  button: {
    paddingHorizontal: 32,
    paddingVertical: 16,
    borderRadius: 8,
    width: '100%',
    alignItems: 'center',
    marginBottom: 12,
  },
  primaryButton: {
    backgroundColor: '#0ea5e9',
  },
  secondaryButton: {
    backgroundColor: 'transparent',
    borderWidth: 2,
    borderColor: '#0ea5e9',
  },
  buttonDisabled: {
    backgroundColor: '#94a3b8',
  },
  buttonText: {
    color: 'white',
    fontSize: 16,
    fontWeight: '600',
  },
  secondaryButtonText: {
    color: '#0ea5e9',
  },
  errorText: {
    color: '#ef4444',
    marginTop: 16,
    textAlign: 'center',
  },
  demoText: {
    color: '#64748b',
    marginTop: 20,
    textAlign: 'center',
    fontSize: 12,
  },
})

export default LoginScreen
